text/0000-style-fn-decl.md

```diff
@@ -0,0 +1,274 @@
+- Feature Name: N/A
+- Start Date: 2015-07-06
+- RFC PR: 
+- Rust Issue: 
+
+# Summary
+
+How should we format function declarations?
+
+# Motivation
+
+Since the last time this was discussed, function declarations have got busier -
+we've added `where` clauses and function types have got larger and are more
+often used in generics. The style I see in the compiler is also different to
+that used in some of the libraries and has moved away from the style guide
+(which is underspecified in any case).
+
+For reference the current text in the style guide states:
+
+> For multiline function signatures, each new line should align with the
+> first parameter. Multiple parameters per line are permitted:
+> 
+> ``` rust
+> fn frobnicate(a: Bar, b: Bar,
+>               c: Bar, d: Bar)
+>               -> Bar {
+>     ...
+> }
+> 
+> fn foo<T: This,
+>        U: That>(
+>        a: Bar,
+>        b: Bar)
+>        -> Baz {
+>     ...
+> }
+> ```
+
+I believe that this does not work very well, since when scanning it is hard to
+distinguish arguments from type parameters and the return type. It also does not
+specify anything for where clauses.
+
+# Detailed design
+
+## Examples
+
+```
+// a comment
+#[an_attribute]
+#[another_attribute]
+pub unsafe extern "C" fn foo<A, B, C>(x: SomeType,
+                                      y: Vec<B>,
+                                      z: SomeOtherType<B, C>)
+    -> ReturnType<A>
+    where A: Foo + Bar,
+          B: Baz<B, C>,
+{
+```
+
+```
+// a comment
+fn foo<A, B, C>(x: SomeType,
+                y: Vec<B>,
+                z: SomeOtherType<B, C>)
+    -> ReturnType<A>
+{
+```
+
+```
+// a comment
+fn foo(x: SomeType,
+       y: Vec<B>,
+       z: SomeOtherType<B, C>)
+{
+```
+
+```
+// a comment
+pub unsafe extern "C" fn foo<A, B, C>(x: SomeType,
+                                      y: Vec<B>,
+                                      z: SomeOtherType<B, C>)
+    where A: Foo + Bar,
+          B: Baz<B, C>,
+{
+```
+
+```
+fn foo<T>(x: T, y: A) -> T where T: Bar {
+
+// alternative
```

Comment:
This alternative is what I personally prefer whenever `fn foo<T>(x: T, y: A) -> T` fits into a single line and `where` part does not.